File: src/Crawler/Analysis/Result/HeadingTreeItem.php
<?php

/*
 * This file is part of the SiteOne Crawler.
 *
 * (c) Ján Regeš <jan.reges@siteone.cz>
 */

declare(strict_types=1);

namespace Crawler\Analysis\Result;

use Crawler\Utils;
use DOMDocument;
use DOMNode;
use DOMXPath;

class HeadingTreeItem
{

    /**
     * Heading level (1-6)
     * @var int
     */
    public readonly int $level;

    /**
     * Real heading level by heading structure in HTML
     * @var int|null
     */
    public ?int $realLevel = null;

    /**
     * Heading text
     * @var string
     */
    public readonly string $text;

    /**
     * Heading ID attribute
     * @var string|null
     */
    public readonly ?string $id;

    /**
     * Parent level just for tree building (it is unset after tree is built)
     * @var HeadingTreeItem|null
     */
    private ?HeadingTreeItem $parent = null;

    /**
     * @var HeadingTreeItem[]
     */
    public array $children = [];

    /**
     * Error text in case of error (typically multiple H1s or wrong heading level)
     * @var string|null
     */
    public ?string $errorText = null;

    /**
     * @param int $level
     * @param string $text
     * @param string|null $id
     */
    public function __construct(int $level, string $text, ?string $id)
    {
        $this->level = $level;
        $this->text = $text;
        $this->id = $id;
    }

    public function addChild(HeadingTreeItem $child): void
    {
        $this->children[] = $child;
    }

    public function hasError(): bool
    {
        return $this->errorText !== null;
    }

    /**
     * @param DOMDocument $dom
     * @param int $maxLevels
     * @return HeadingTreeItem[]
     */
    public static function getHeadingTreeFromHtml(DOMDocument $dom, int $maxLevels = 3): array
    {
        $xpath = new DOMXPath($dom);
        $xPathAppend = '';
        for ($i = 2; $i <= $maxLevels; $i++) {
            $xPathAppend .= " or self::h{$i}";
        }
        $nodes = @$xpath->query('//*[self::h1' . $xPathAppend . ']');

        $root = new HeadingTreeItem(0, '', null);
        $currentNode = $root;

        $h1References = [];

        foreach ($nodes ?: [] as $node) {
            /* @var $node DOMNode */
            $level = (int)substr($node->nodeName, 1);

            // WARNING: $node->textContent is not working properly in cases where the website uses other HTML elements
            // inside <h*>, including <script>, so JS code (without <script> tag) is included in the textContent
            //  $text = trim(preg_replace('/\s+/', ' ', $node->textContent));
            $nodeContent = strip_tags(Utils::stripJavaScript($node->ownerDocument->saveHTML($node)));
            $text = trim(preg_replace('/\s+/', ' ', $nodeContent));

            $id = $node->getAttribute('id') ?: null;

            $item = new HeadingTreeItem($level, $text, $id);
            if ($level === 1) {
                $h1References[] = $item;
            }

            while ($currentNode->level >= $level) {
                $currentNode = $currentNode->parent;
            }

            $item->parent = $currentNode;
            $currentNode->children[] = $item;

            $currentNode = $item;
        }

        // set error to multiple h1s if exists
        $h1Count = $dom->getElementsByTagName('h1')->length;
        if ($h1Count > 1) {
            foreach ($h1References as $h1Reference) {
                $h1Reference->errorText = "Multiple H1s ({$h1Count}) found.";
            }
        }

        $finalTreeItemFixes = function (HeadingTreeItem $item, int $realLevel = 1) use (&$finalTreeItemFixes) {
            unset($item->parent);
            $item->realLevel = $realLevel;
            foreach ($item->children as $child) {
                unset($child->parent);
                $finalTreeItemFixes($child, $realLevel + 1);
            }

            if ($item->level !== $item->realLevel) {
                $item->errorText = "Heading level {$item->level} is not correct. Should be {$item->realLevel}.";
            }
        };

        $result = array_filter($root->children, fn($child) => $child instanceof HeadingTreeItem);
        foreach ($result as $item) {
            $finalTreeItemFixes($item);
        }
        return $result;
    }

    /**
     * @param HeadingTreeItem[] $items
     * @return string
     */
    public static function getHeadingTreeUlLiList(array $items): string
    {
        $result = '<ul>';
        foreach ($items as $item) {
            $result .= '<li>' . self::getHeadingTreeUlLi($item, true) . '</li>';
        }
        $result .= '</ul>';
        return $result;
    }

    public static function getHeadingTreeUlLi(HeadingTreeItem $item, bool $addItem = true): string
    {
        $result = '';
        if ($addItem) {
            $txtRow = "<h{$item->level}> {$item->text}" . ($item->id ? ' [#' . $item->id . ']' : '');
            $result = $item->hasError()
                ? (
                    '<span class="help" title="' . htmlspecialchars($item->errorText) . '">'
                    . Utils::getColorText(htmlspecialchars($txtRow), 'magenta')
                    . '</span>'
                )
                : htmlspecialchars($txtRow);
        }

        if ($item->children) {
            $result .= '<ul>';
            foreach ($item->children as $child) {
                $result .= '<li>';
                $txtRow = "<h{$child->level}> " . $child->text . ($child->id ? ' [#' . $child->id . ']' : '');
                $result .= $child->hasError()
                    ? (
                        '<span class="help" title="' . htmlspecialchars($child->errorText) . '">'
                        . Utils::getColorText(htmlspecialchars($txtRow), 'magenta')
                        . '</span>'
                    )
                    : htmlspecialchars($txtRow);
                $result .= self::getHeadingTreeUlLi($child, false);
                $result .= '</li>';
            }
            $result .= '</ul>';
        }
        return $result;
    }

    /**
     * @param HeadingTreeItem[] $items
     * @return string
     */
    public static function getHeadingTreeTxtList(array $items): string
    {
        $result = '';
        foreach ($items as $item) {
            $result .= self::getHeadingTreeTxt($item, true);
        }
        return preg_replace('/\s+/', ' ', $result);
    }

    public static function getHeadingTreeTxt(HeadingTreeItem $item, bool $addItem = true): string
    {
        $result = '';
        if ($addItem) {
            $result .= "<h{$item->level}> {$item->text}" . ($item->id ? ' [#' . $item->id . ']' : '') . "\n";
        }
        foreach ($item->children as $child) {
            $result .= str_repeat('  ', $child->level - 1);
            $txtRow = "<h{$child->level}> {$child->text}" . ($child->id ? ' [#' . $child->id . ']' : '');
            // $result .= ($child->level !== $child->realLevel ? Utils::getColorText($txtRow, 'magenta') : $txtRow);
            $result .= $txtRow;
            $result .= "\n";
            $result .= self::getHeadingTreeTxt($child, false);
        }
        return $result;
    }

}

File: src/analysis/result/heading_tree_item.rs
// SiteOne Crawler - HeadingTreeItem
// (c) Jan Reges <jan.reges@siteone.cz>

fn html_escape(s: &str) -> String {
    s.replace('&', "&amp;")
        .replace('<', "&lt;")
        .replace('>', "&gt;")
        .replace('"', "&quot;")
        .replace('\'', "&#39;")
}

#[derive(Debug, Clone)]
pub struct HeadingTreeItem {
    /// Heading level (1-6)
    pub level: i32,
    /// Real heading level by heading structure in HTML
    pub real_level: Option<i32>,
    /// Heading text
    pub text: String,
    /// Heading ID attribute
    pub id: Option<String>,
    /// Children headings
    pub children: Vec<HeadingTreeItem>,
    /// Error text in case of error (typically multiple H1s or wrong heading level)
    pub error_text: Option<String>,
}

impl HeadingTreeItem {
    pub fn new(level: i32, text: String, id: Option<String>) -> Self {
        Self {
            level,
            real_level: None,
            text,
            id,
            children: Vec::new(),
            error_text: None,
        }
    }

    pub fn has_error(&self) -> bool {
        self.error_text.is_some()
    }

    /// Get heading tree as a plain text list
    pub fn get_heading_tree_txt_list(items: &[HeadingTreeItem]) -> String {
        let mut result = String::new();
        for item in items {
            result.push_str(&Self::get_heading_tree_txt(item, true));
        }
        // Collapse whitespace
        let re = regex::Regex::new(r"\s+").unwrap_or_else(|_| regex::Regex::new(".^").unwrap());
        re.replace_all(&result, " ").trim().to_string()
    }

    fn get_heading_tree_txt(item: &HeadingTreeItem, add_item: bool) -> String {
        let mut result = String::new();
        if add_item {
            result.push_str(&format!("<h{}> {}", item.level, item.text));
            if let Some(ref id) = item.id {
                result.push_str(&format!(" [#{}]", id));
            }
            result.push('\n');
        }
        for child in &item.children {
            result.push_str(&"  ".repeat((child.level - 1) as usize));
            result.push_str(&format!("<h{}> {}", child.level, child.text));
            if let Some(ref id) = child.id {
                result.push_str(&format!(" [#{}]", id));
            }
            result.push('\n');
            result.push_str(&Self::get_heading_tree_txt(child, false));
        }
        result
    }

    /// Get heading tree as an HTML `<ul><li>` list.
    pub fn get_heading_tree_ul_li_list(items: &[HeadingTreeItem]) -> String {
        let mut result = String::from("<ul>");
        for item in items {
            result.push_str("<li>");
            result.push_str(&Self::get_heading_tree_ul_li(item, true));
            result.push_str("</li>");
        }
        result.push_str("</ul>");
        result
    }

    fn get_heading_tree_ul_li(item: &HeadingTreeItem, add_item: bool) -> String {
        let mut result = String::new();
        if add_item {
            let txt_row = format!(
                "&lt;h{}&gt; {}{}",
                item.level,
                html_escape(&item.text),
                item.id
                    .as_ref()
                    .map(|id| format!(" [#{}]", html_escape(id)))
                    .unwrap_or_default()
            );
            if item.has_error() {
                let error_text = html_escape(item.error_text.as_deref().unwrap_or(""));
                let colored = crate::utils::get_color_text(&txt_row, "magenta", false);
                let colored_html = crate::utils::convert_bash_colors_in_text_to_html(&colored);
                result.push_str(&format!(
                    "<span class=\"help\" title=\"{}\">{}</span>",
                    error_text, colored_html
                ));
            } else {
                result.push_str(&txt_row);
            }
        }

        if !item.children.is_empty() {
            result.push_str("<ul>");
            for child in &item.children {
                result.push_str("<li>");
                let txt_row = format!(
                    "&lt;h{}&gt; {}{}",
                    child.level,
                    html_escape(&child.text),
                    child
                        .id
                        .as_ref()
                        .map(|id| format!(" [#{}]", html_escape(id)))
                        .unwrap_or_default()
                );
                if child.has_error() {
                    let error_text = html_escape(child.error_text.as_deref().unwrap_or(""));
                    let colored = crate::utils::get_color_text(&txt_row, "magenta", false);
                    let colored_html = crate::utils::convert_bash_colors_in_text_to_html(&colored);
                    result.push_str(&format!(
                        "<span class=\"help\" title=\"{}\">{}</span>",
                        error_text, colored_html
                    ));
                } else {
                    result.push_str(&txt_row);
                }
                result.push_str(&Self::get_heading_tree_ul_li(child, false));
                result.push_str("</li>");
            }
            result.push_str("</ul>");
        }
        result
    }

    /// Count total headings in tree
    pub fn get_headings_count(items: &[HeadingTreeItem]) -> usize {
        let mut count = 0;
        for item in items {
            count += 1;
            count += Self::get_headings_count(&item.children);
        }
        count
    }

    /// Count headings with errors in tree
    pub fn get_headings_with_error_count(items: &[HeadingTreeItem]) -> usize {
        let mut count = 0;
        for item in items {
            if item.has_error() {
                count += 1;
            }
            count += Self::get_headings_with_error_count(&item.children);
        }
        count
    }
}
